Feat: add react final task en (#781)
rolling-scopes-school/tasks: tasks/react/project-management-system-EN.md

#  Project Management System

**Project management system** is an application that helps an individual in a team or group of developers achieve their goals.

## Theoretical Notes

> There are many competitors on the market for our future application. Although this is not scary, we decided to research
> them before making our own. The marketing department figured out the main competitors are the following: *Trello, Jira, Redmine, Bitrix24, 
> Yandex Tracker, Asana, GanttPro, Github projects.*

Reviewer: ```suggestion
There are many competitors on the market for our future application. Check out suggested analogs before start building your app: *Trello, Jira, Redmine, Bitrix24, Yandex Tracker, Asana, GanttPro, Github projects.*
```

## Application Prototype

Pay attention that we leave the final set of tools and design up to you to not limit your possibilities and imagination.

Design, prototype, as well as the implementation of the application itself to your taste.

Here's the list of examples of working prototypes:

https://comfy-klepon-a13665.netlify.app/

https://project-management-appl.herokuapp.com/

https://pma-angular.netlify.app/

https://project-management-app-group14.netlify.app/

https://project-management-team7.netlify.app/

## Backend

> Use https://github.com/vitaly-sazonov/kanban-rest or https://github.com/rolling-scopes-school/final-task-backend/pull/1 as a backend.

Reviewer: ```suggestion
Use https://github.com/vitaly-sazonov/kanban-rest as a backend.
```

There is a description of the steps to get started, as well as a list of available requests in the repository.

- You will need to deploy the backend yourself to demonstrate the working application.
- During the implementation of an application, you may run the backend in any environment that is convenient for you, for example, a local machine.

## Application structure

> Regardless of the application type, it should contain:

Reviewer: ```suggestion
Your app must contain:
```

- welcome page

> - user login

Reviewer: ```suggestion
- user auth
```

- projects listing page
- selected project management page

> - additional functionality (for example, the ability to view all the tasks of the selected user)

Reviewer: ```suggestion
Your app could contain:
- additional functionality (for example, the ability to view all the tasks of the selected user)
```

## Repository requirements

> - use your private school repository for this task

Reviewer: ```suggestion
- teamlead should create private repo in github and invite other members 
```

> - branch name is **project-management-app**

Reviewer: ```suggestion
- repo name **project-management-app**. Branch name for develop - **develop**, **main** branch is empty and contains only README.md
```

> - the commit history should reflect the development process of the application. [Commit requirements](https://docs.rs.school/#/git-convention)

Reviewer: ```suggestion
- the commit history should display application development process. [Commit requirements](https://docs.rs.school/#/git-convention)
```

- the demo version of the application is hosted on `gh-pages`, `netlify`, `heroku`, or another similar hosting.

> - after completing the task, create a pull request **Merging a Pull Request is not required**

Reviewer: ```suggestion
- after completing the task, create a pull request from **development** to **master** branch. **Merging a Pull Request is not required**
```

> - add a link to the backend folder to the description of the pull request if the backend has not been deployed.

Reviewer: ```suggestion
- describe in PR how to run your backend locally if you didn't deploy backend.
```

## Technical requirements

> - localization (at least 2 languages). The application should be able to change the language by clicking on the slider in the header

Reviewer: ```suggestion
- localization (at least 2 languages). You should be able to change the language by clicking on the toggler/select in header.
```

- adaptive layout. The minimum page width of the application is 320px
- interactivity of elements users can interact with, changing the appearance of the element itself and the type of the cursor on hover, using different styles for the active and inactive state of the element, smooth animations
- the unity of styles of all pages of the application - the same fonts, button styles, indents, the same elements on all pages of the application have the same appearance and layout. Item colors and background images may vary. In this case, colors should be from the same palette, and background images from the same collection.

## Description of function blocks

### Welcome page(route)

- The welcome page should display general information about the developer, project, and course.
- In the upper right corner should be present 2 buttons Login and Sign up
- If there is an unexpired token, the user should be redirected to the "Main route" of the application automatically.
- When the token expires - the user should be redirected to the "Welcome page" automatically.
- Pressing the Login / Sign up button redirects the user to the route with the Login / Sign up form automatically.

> ### Header

Reviewer: ```suggestion
### Header

- All Privite routes should contain sticky header. Moment when it become sticky (if there is a scroll on a page) should be animated: color can be changed or it's height can become smaller. [Animated sticky header](https://www.youtube.com/watch?v=hR8UW5CvYgw)
```

- There are buttons in the header for authenticated users: edit profile, logout, create new board, and localization switcher.

> - Edit profile should redirect the user to a route with a form for edit profile. The requirements for the form are the same as for all forms in the application. There should be a 'Delete User' button. In case of this action should be shown as a  "confirmation modal" then the user should be logged out, and the user should be removed from the database.

Reviewer: ```suggestion
- Edit profile button should redirect the user to a route with a form for edit profile. The requirements for the form are the same as for all forms in the application. There should be a 'Delete User' button. In case of this action => "confirmation modal" => then the user should be logged out, and the user should be removed from the database.
```

- Create a new board - opens a modal window with a form for creating a board. Requirements for the form are the same as for all forms in the application.

### Footer

- footer should contain a link to the author's github, the year the application was created, [course logo](https://rs.school/images/rs_school_js.svg) with [link to the course](https://rs.school/react/).
- footer is displayed on all pages of the application.
### Login / Sign up

- Form fields should be implemented in consistency with the backend API of the application. Validation should be implemented.
- Errors from the BE side - (Not found, unhandled rejection, etc) should be displayed in a user-friendly format (toast, pop-up, or something like this - up to your decision).

> - Upon successful login, the user should be redirected to the "Main route"

Reviewer: ```suggestion
- Upon successful login, the user should be redirected to the "Main route"
- If user already logged in and he try to reach this routes - he should be redirected to Main route.
```

- ATTENTION! Deleting a column removes the tasks associated with it from the BD automatically.
- There should be a "back" button to return to the main route

## Evaluation criteria

**Maximum available points for the task 620**

## Check criteria

### Welcome route - max 70 points

- [ ] The welcome page should contain general information about the developer, project, and course. **10 points**
- [ ] In the upper right corner there are 2 buttons: login and sign up. **10 points**
- [ ] If there is an unexpired token, the user should be redirected to the "Main route" of the application automatically. **20 points**
- [ ] When the token expires - the user should be redirected to the "Welcome page" automatically. **20 points**
- [ ] Pressing the Login / Sign up button redirects a user to the route with the Login / Sign up form. **10 points**

### Login / Sign up  - max 80 points

- [ ] Login/log out should be present on all pages **20 points**
- [ ] Form fields should be implemented according to the backend API. Validation should be implemented. **50 points**
- [ ] Upon successful login, the user should be redirected to "Main route" **10 points**

> ### Main route - max 100 points
> 
> - [ ] Board creation functionality **20 points**
> - [ ] Displays all created boards as a list **10 points**
> - [ ] Each board in the list is displayed with a small preview of available information (title, description, etc). By clicking an element the user navigates to the board item (Board route). There's also a button for board deletion. **10 points**
> - [ ] When trying to delete the board, we should receive a confirmation modal. The confirmation modal must be a generic component (one for the entire application). **10 points**
> - [ ] Global search: search for a task by a task number, name, users who participate in it, and by the text of the task description. **20 points**
> - [ ] The user profile editing functionality is implemented. **30 points**

Reviewer: ```suggestion
### Main route - max 8 points

- [ ] Board creation functionality **2 points**
- [ ] Displays all created boards as a list/grid **1 point**
- [ ] Each board in the list is displayed with a small preview of available information (title, description, etc). By clicking an element the user navigates to the board item (Board route). There's also a button for board deletion. **1 point**
- [ ] When trying to delete the board, we should receive a confirmation modal. The confirmation modal must be a generic component (one for the entire application). **1 points**
- [ ] The user profile editing functionality is implemented. **3 points**
```

> ### Board route - max 260 points
> 
> - [ ] Button for column creation is displayed **10 points**
> - [ ] If a board contains at least one column - a button for task creation is displayed as well **10 points**
> - [ ] A modal window with form is displayed for column and task creation  **30 points**
> - [ ] A vertical scrollbar is displayed in the column when overflowing with the number of column tasks  **20 points**
> - [ ] The page itself on the current route doesn't have a vertical scrollbar **10 points**
> - [ ] With the help of drag-n-drop, we can swap columns. **30 points**
> - [ ] With the help of drag-n-drop, we can change the order of tasks within a column.  **30 points**
> - [ ] With the help of drag-n-drop, we can change the task belonging to the column.  **50 points**
> - [ ] by clicking on the task, we open a modal window with the edit task form. The requirements for the form and window are the same as everywhere else. **30 points**
> - [ ] The task must have a delete task button. On click: confirmation modal -> delete.  **10 points**
> - [ ] The top of a column should always display the column title. By clicking the title the test should turn into a form with cancel and submit buttons. After typing a text into the input and clicking the submit button the tile of the column should be updated. **20 points**
> - [ ] The column should have a delete button. By clicking -> confirmation modal -> when approving -> deleting. **10 points**

Reviewer: ```suggestion
### Board route - max 26 points

- [ ] Button for column creation is displayed **1 point**
- [ ] If a board contains at least one column - a button for task creation is displayed/become enabled as well **1 points**
- [ ] A modal windows with forms is displayed for column and task creations  **3 points**
- [ ] A vertical scrollbar is displayed in the column when overflowing with the number of column tasks  **2 points**
- [ ] The page itself on the current route doesn't have a vertical scrollbar **1 points**
- [ ] With the help of drag-n-drop, we can swap columns. **3 points**
- [ ] With the help of drag-n-drop, we can change the order of tasks within a column.  **3 points**
- [ ] With the help of drag-n-drop, we can change the task belonging to the column.  **5 points**
- [ ] The functionality of viewing and editing of the task has been implemented. **3 points**
- [ ] The task must have a delete task button. On click: confirmation modal -> delete.  **1 points**
- [ ] At the top of the column should be Title. When you click on it, it should become an input, with Submit and Cancel buttons near it. After entering text in the input and clicking Submit - the Title of the column should change. **2 points**
- [ ] The column should have a delete button. By clicking -> confirmation modal -> when approving -> deleting. **1 points**
```

> ### General requirements - max 80 points
> 
> - [ ] Backend error handling - (Not found, unhandled rejection, etc) should be performed in a user-friendly way (toast, pop-up or anything else you implement). **50 points**
> - [ ] Localization **20 points**
> - [ ] Backend is deployed and publicly available **10 points**

Reviewer: ```suggestion
### General requirements - max 11 points

- [ ] Backend error handling - (Not found, unhandled rejection, etc) should be performed in a user-friendly way (toast, pop-up or anything else you implement). **4 points**
- [ ] Localization **1 point**
- [ ] Backend is deployed and publicly available **2 points**
- [ ] Sticky header **2 points**
- [ ] Extra scope same complexity as Global search **2 points**
```

> ### Additiona functionality - 30 points
> 
> - [ ] Implemented additional functionality that is not provided for by current requirements **30 points**

Reviewer: ```suggestion

```

> ### Penalties
> 
> - [ ] The presence of errors and warnings in the console  **- 20 points**
> - [ ] The presence in the console of the results of the console.log execution - **- 20 points**
> - [ ] Making commits after the deadline - **- 100 points**

Reviewer: ```suggestion
### Penalties
- [ ] React default favicon **- 1 points**
- [ ] The presence of errors and warnings in the console  **- 2 points** for each
- [ ] The presence in the console of the results of the console.log execution - **- 2 points** for each
- [ ] @ts-ignore or any usage (search through github repo) - **- 1 point** for each
- [ ] Making commits after the deadline - **- 20 points**  

## Helpful links

- [Project management](https://ru.wikipedia.org/wiki/%D0%A3%D0%BF%D1%80%D0%B0%D0%B2%D0%BB%D0%B5%D0%BD%D0%B8%D0%B5_%D0%BF%D1%80%D0%BE%D0%B5%D0%BA%D1%82%D0%B0%D0%BC%D0%B8)
- [What is a project?](https://ru.wikipedia.org/wiki/%D0%9F%D1%80%D0%BE%D0%B5%D0%BA%D1%82_(%D0%B2_%D1%83%D0%BF%D1%80%D0%B0%D0%B2%D0%BB%D0%B5%D0%BD%D1%87%D0%B5%D1%81%D0%BA%D0%BE%D0%B9_%D0%B4%D0%B5%D1%8F%D1%82%D0%B5%D0%BB%D1%8C%D0%BD%D0%BE%D1%81%D1%82%D0%B8))
- [Kanban](https://en.wikipedia.org/wiki/Kanban_(development))
```
